makefile: allow users to define a dockercmd

raghavendra-talur opened the pull request:
This allows one to use a different tool, say podman instead of docker for their container build.

Signed-off-by: Raghavendra Talur <raghavendra.talur@gmail.com>

nirs commented on Makefile:
```diff
@@ -165,10 +172,10 @@ run-dr-cluster: generate manifests ## Run DR manager controller from your host.
 
 docker-build: ## Build docker image with the manager.
-	docker build -t ${IMG} .
+	$(DOCKERCMD) build -t ${IMG} .
```
Maybe CONTAINER_TOOL?

nirs commented on Makefile:
```diff
@@ -84,6 +84,13 @@ endif
+ifeq ($(DOCKERCMD),)
+DOCKERCMD=$(shell podman version >/dev/null 2>&1 && echo podman)
+endif
+endif
```
Podman is available everywhere, and is a drop-in replacement for docker, so we can
safely use podman instead of docker and document that we require it.

If we want to allow using docker instead of podman we can provide a environment
variable like CONTAINER_TOOL and let the user override it without any checks.

If the tool is not available the command will fail in a pretty clear way:

    $ docker build -t foo .
    bash: docker: command not found...

Maintaining cryptic makefiles is not fun, better to keep it simple.

nirs replied:
Being able to use `alias docker=podman` is nice, but it is a hack and I don't want hacks
in my environment. I want to know exactly what is running and be able to run both podman
and docker if needed. Installing podman-docker is a better hack also handling man pages
and doing system level change cleanly.

But do have an issue switching to podman? This should affect only ramen developers.

ShyamsundarR replied:
I really do not have an issue, it's just that operator sdk devs (who visit/contribute in the future) would have something new to deal with, which we can avoid. I have seen this DOCKERCMD or variants in a few Makefiles on other projects, so not a blocker from my end, just stating there are alternatives out there.

ShyamsundarR replied:
Looking at this a bit more, "undockerizing" this for the env variable is only a start, as we have various docker* makefile targets as well. As a result we can just retain a simple environment variable (say DOCKERCMD) with it's default as `podman` and let users override this in their shell as required. (As @nirs states keeps the Makefile a little simpler)

raghavendra-talur replied:
Updated the patch to default to podman.

nirs reviewed (commented):
Nice, I don't want docker on my machines :-)

nirs commented on Makefile:
```diff
@@ -87,6 +87,10 @@ endif
 
+ifeq ($(DOCKERCMD),)
+DOCKERCMD = podman
+endif
```
Why not use the same method used for other variables?

    DOCKERCMD ?= podman

See https://www.gnu.org/software/make/manual/make.html#Setting

raghavendra-talur replied:
Thanks, made the update now.

ShyamsundarR commented:
@raghavendra-talur the GH action uses docker, so either we need to update that to podman or we need to add the DOCKERCMD env to point to docker in sanity.yaml file (the checks are failing due to the same)

raghavendra-talur commented:
@ShyamsundarR I fixed the github workflow error.

I have also split the jobs for the github workflow to run tests on their own. This also makes the tests a bit faster as the unit tests and the image build jobs run in parallel.

ShyamsundarR commented on .github/workflows/sanity.yml:
```diff
@@ -138,8 +163,7 @@ jobs:
         run: |
-          docker load -i /tmp/ramen-operator.tar
-          kind load docker-image ${IMAGE_TAG_BASE}-operator:${IMAGE_TAG} --name ${KIND_CLUSTER_NAME}
+          kind load image-archive /tmp/ramen-operator.tar --name ${KIND_CLUSTER_NAME}
```
(nit) Why in this one instance it is `kind` and not `${{env.DOCKERCMD}}`?

raghavendra-talur replied:
Well, I just deleted the docker line above. Earlier, it was a 2 step process where docker loaded the tar into its images and then kind loaded the image from docker daemon to the the cluster.

If we are using podman, with the version of kind we are using, it is not possible to use the `kind load docker-image` variant(it requires docker daemon and podman isn't sufficient).

Instead, we load the image into the cluster in a single step by using the `kind load image-archive` command that takes the tar variant directly.

ShyamsundarR commented on .github/workflows/sanity.yml:
```diff
@@ -194,7 +218,7 @@ jobs:
 
```
`publish-image` job needs to depend on the `unit tests` and `ensure clean branch` checks, so that we publish onyl if all jobs are successful (when building from main on a merge)

raghavendra-talur replied:
Done
